### PR TITLE
Add AnnotationAttributeDualReader that can read both annotations and attributes at the same time

### PR DESCRIPTION
**Pull request by TomasVotruba**
I'm testing the practise in practise and it seems the `DualReader` can only read annotation or attributes, not both at the same time.

That's why method like this fails:

```php

	/**
	 * @RequestParameters({
	 *     @RequestParameter(in="path", type="int", name="userId", description="User ID"),
	 * })
	 */
	#[Path(path: '/user/{userId}/collections')]
	public function run()
	{
	}
```

### CODE REVIEW

File: src/MetadataReader/AnnotationAttributeDualReader.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+use ReflectionClass;
+use ReflectionMethod;
+use ReflectionProperty;
+
+/**
+ * This reader reads BOTH annotations and attributes above one elements at the same time,
+ * and merge them together.
+ * Compared to @see DualReader that only prefers one or the other.
+ */
+final class AnnotationAttributeDualReader implements Reader
```
Comment (TomasVotruba) [inline src/MetadataReader/AnnotationAttributeDualReader.php:17]: @koriym Hi, I came to a bug when only annotations or attributes are used. The `DualReader` can handle only one of them, depending on PHP version.

This caused missing annotations on PHP 8. I think more reliable solution would be class like this one - that reads both annotations and attributes on PHP 8, if available.  That's what I expected from `DualReader` class and that's where it failed my expectations. But maybe there is a better way to do it.

What do you think?

**Comment by TomasVotruba** [comment]
Ready to merge :+1:

File: src/MetadataReader/AnnotationAttributeDualReader.php (hunk 1 above)
Comment (koriym) [reply src/MetadataReader/AnnotationAttributeDualReader.php:17]: As you pointed out! It's a bug and needs to be fixed.

Comment (TomasVotruba) [reply src/MetadataReader/AnnotationAttributeDualReader.php:17]: Yes, exactly. It fails for the arrays

File: src/Annotation/Controller/RequestParameters.php [hunk 2]
```diff
@@ -31,6 +31,24 @@ public function __construct($parameters)
...
+	public function getParameters(): array
+	{
+		return $this->parameters;
+	}
+
+	/**
+	 * @param RequestParameter[] $parameters
+	 */
+	private function validateUniqueNames(array $parameters): void
+	{
```
Comment (TomasVotruba) [inline src/Annotation/Controller/RequestParameters.php:51]: Decoupling complex validation code that does not modify/work with values makes the class more readable

File: src/MetadataReader/AnnotationAttributeDualReader.php (hunk 1 above)
Comment (koriym) [reply src/MetadataReader/AnnotationAttributeDualReader.php:17]: Issue created. https://github.com/koriym/Koriym.Attributes/issues/10

**Comment by TomasVotruba** [comment]
There is one more bug with doctrine/annotations... :thinking:  I'm just looking into it

File: tests/fixtures/Controllers/Mixed/PathAndRequestParamsController.php [hunk 3]
```diff
@@ -13,10 +13,10 @@ final class PathAndRequestParamsController extends ApiV1Controller
 {
 
 	/**
-	 * @RequestParameters({
+	 * @RequestParameters(
 	 *     @RequestParameter(name="userId", type="int", description="User ID", in="path"),
 	 *     @RequestParameter(name="photoId", type="int", description="Verification photo ID", in="path")
-	 * })
+	 * )
 	 */
```
Comment (TomasVotruba) [inline tests/fixtures/Controllers/Mixed/PathAndRequestParamsController.php:20]: Reported in https://github.com/doctrine/annotations/issues/418

File: composer.json [hunk 4]
```diff
@@ -23,7 +23,7 @@
     "ext-json": "*",
     "contributte/psr7-http-message": "~0.7.0",
     "doctrine/annotations": "^1.13",
-    "koriym/attributes": "^1.0",
+    "koriym/attributes": "1.x-dev#2b0349d",
     "nette/utils": "^3.0.3"
   },
```
Comment (TomasVotruba) [inline composer.json:28]: @koriym Btw, I'm testing https://github.com/koriym/Koriym.Attributes/pull/11 here and everything passes well :+1: 

I've removed the custom DualReader implementation from here and used the native one.

File: tests/cases/DI/Loader/MixedAttributeAnnotationLoader.phpt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	$loader = new DoctrineAnnotationLoader($builder);
+	$schemaBuilder = $loader->load(new SchemaBuilder());
+	Assert::type(SchemaBuilder::class, $schemaBuilder);
+
+	$controllers = $schemaBuilder->getControllers();
+	Assert::count(3, $controllers);
+
+	$requestParameterValidation = new RequestParameterValidation();
+	$requestParameterValidation->validate($schemaBuilder);
+});
```
Comment (TomasVotruba) [inline tests/cases/DI/Loader/MixedAttributeAnnotationLoader.phpt:41]: This PR does not add any behavior change.
It only **adds 3 tests cases that should be all validated successfuly** (based on fails in real project :)).

It containts 3 combinations:

* attributes only :heavy_check_mark: 
* one attribute and annotation :heavy_check_mark: 
* two attribute and annotation :heavy_check_mark:

File: composer.json [hunk 6]
```diff
@@ -23,7 +23,7 @@
     "ext-json": "*",
     "contributte/psr7-http-message": "~0.7.0",
     "doctrine/annotations": "^1.13",
-    "koriym/attributes": "^1.0",
+    "koriym/attributes": "1.x-dev#2b0349d",
```
Comment (f3l1x) [inline composer.json:26]: Do you know when will be release stable version? 1.1 or 1.0.1?

Comment (TomasVotruba) [reply composer.json:26]: Working on it https://github.com/koriym/Koriym.Attributes/pull/11#issuecomment-878144758 :)

**Comment by TomasVotruba** [comment]
I've updated version to use tag, now up and CI running succesfuly :tada: 

Ready to merge :heavy_check_mark:

**Comment by TomasVotruba** [comment]
Just curious, our project avoid dev deps to speedup compose installations on CI and make dependencies more reliable. When do you plan to release this feature?

**Comment by f3l1x** [comment]
I will call you on the phone. ;-)

**Pull request merged**